dask_cuda/explicit_comms/dataframe/shuffle.py

```diff
@@ -5,201 +5,239 @@
-    except AttributeError:
-        manager = None
-
-    if isinstance(manager, ProxyManager):
-
-        def concat(args, ignore_index=False):
-            if len(args) < 2:
-                return args[0]
-
-            return manager.proxify(dd_concat(args, ignore_index=ignore_index))[0]
-
-    else:
-        concat = dd_concat
-
-    rank_to_out_parts_list = sort_in_parts(
-        in_parts,
-        rank_to_out_part_ids,
-        ignore_index,
-        concat_dfs_of_same_output_partition=True,
-        concat=concat,
+    stage: dict = s["stages"].pop(stage_name)
+    assert stage.keys() == rank_to_inkeys[myrank]
+
+    out_part_id_to_dataframe = multi_shuffle_group(
+        df_meta=df_meta,
+        dfs=stage,
+        column_names=column_names,
+        npartitions=npartitions,
+        ignore_index=ignore_index,
+        proxify=proxify,
     )
 
     # Communicate all the dataframe-partitions all-to-all. The result is
-    # `out_parts_list` that for each worker and for each output partition
-    # contains a list of dataframes received.
-    out_parts_list: List[List[List[DataFrame]]] = []
-    futures = []
-    if myrank in rank_to_out_parts_list:
-        futures.append(recv(eps, in_nparts, out_parts_list))
-    if myrank in in_nparts:
-        futures.append(send(eps, rank_to_out_parts_list))
-    await asyncio.gather(*futures)
+    # `out_part_id_to_dataframe_list` that for each output partition maps
+    # a list of dataframes received.
+    out_part_id_to_dataframe_list: Dict[int, List[DataFrame]] = defaultdict(list)
+    await asyncio.gather(
+        recv(eps, myrank, rank_to_out_part_ids, out_part_id_to_dataframe_list, proxify),
+        send(eps, myrank, rank_to_out_part_ids, out_part_id_to_dataframe),
+    )
 
     # At this point `send()` should have pop'ed all output partitions
     # beside the partitions owned be `myrank`.
-    assert len(rank_to_out_parts_list) == 1
-
-    # Concatenate the received dataframes into the final output partitions
-    ret = []
-    for i in range(len(rank_to_out_part_ids[myrank])):
-        dfs = []
-        for out_parts in out_parts_list:
-            dfs.extend(out_parts[i])
-            out_parts[i] = None  # type: ignore
-        dfs.extend(rank_to_out_parts_list[myrank][i])
-        rank_to_out_parts_list[myrank][i] = None  # type: ignore
-        if len(dfs) > 1:
-            ret.append(concat(dfs, ignore_index=ignore_index))
-        else:
-            ret.append(dfs[0])
+    assert rank_to_out_part_ids[myrank] == out_part_id_to_dataframe.keys()
+    # We can now add them to the output dataframes.
+    for out_part_id, dataframe in out_part_id_to_dataframe.items():
+        out_part_id_to_dataframe_list[out_part_id].append(dataframe)
+    del out_part_id_to_dataframe
+
+    # Finally, we concatenate the output dataframes into the final output partitions
+    ret: List[DataFrame] = []
+    for out_part_id, dataframe_list in out_part_id_to_dataframe_list.items():
```

Comment:
```suggestion
    for dataframe_list in out_part_id_to_dataframe_list.values():
```

You never needed the id.

So probably a list comprehension here would be simpler:

```
ret = [
proxify(dd_concat(dfs, ignore_index=ignore_index)) for dfs in out_part_id_to_dataframe_list.values()
]
```

Reply:
Good point, fixed in 6464eb6b9f3b07a11276729eb2a25d37149377f7